File: fastapi_server/fastapi_server/bootstrap/registry/registry.py
from __future__ import annotations

import inspect
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Dict, List, Union

from fastapi_server.bootstrap.contract import (
    LoaderReport,
    ResolvedBootstrapConfig,
    create_loader_report,
    report_error,
)

RunFn = Callable[[Any, ResolvedBootstrapConfig, Any], Union[LoaderReport, Awaitable[LoaderReport]]]


@dataclass
class Addon:
    name: str
    priority: int
    run: RunFn


@dataclass
class _Entry:
    addon: Addon
    insertion_index: int


class Registry:
    def __init__(self) -> None:
        self._entries: List[_Entry] = []
        self._counter = 0

    def register(self, addon: Addon) -> "Registry":
        if not isinstance(addon, Addon):
            raise TypeError("expected Addon instance")
        if not callable(addon.run):
            raise TypeError("addon.run must be callable")
        self._entries.append(_Entry(addon=addon, insertion_index=self._counter))
        self._counter += 1
        return self

    def list(self) -> List[Addon]:
        return [
            e.addon
            for e in sorted(self._entries, key=lambda e: (e.addon.priority, e.insertion_index))
        ]

    async def run_all(
        self,
        server: Any,
        config: ResolvedBootstrapConfig,
        ctx: Any,
    ) -> Dict[str, LoaderReport]:
        reports: Dict[str, LoaderReport] = {}
        for addon in self.list():
            try:
                result = addon.run(server, config, ctx)
                if inspect.isawaitable(result):
                    result = await result
                reports[addon.name] = result  # type: ignore[assignment]
            except Exception as err:  # noqa: BLE001
                fallback = create_loader_report(addon.name)
                report_error(fallback, "run", err)
                reports[addon.name] = fallback
        return reports


def compose_addons(*groups: Union[Addon, List[Addon]]) -> List[Addon]:
    out: List[Addon] = []
    for g in groups:
        if isinstance(g, list):
            out.extend(g)
        else:
            out.append(g)
    return out

File: fastify_server/src/registry/registry.ts
import type { LoaderReport, LoaderReports, ResolvedBootstrapConfig } from "../contract/index.js";
import { createLoaderReport, reportError } from "../contract/index.js";
import type { RuntimeContext } from "./ctx.js";

export interface Addon<Server = unknown> {
  name: string;
  priority: number;
  run: (
    server: Server,
    config: ResolvedBootstrapConfig,
    ctx: RuntimeContext<Server>,
  ) => LoaderReport | Promise<LoaderReport>;
}

interface RegistryEntry<Server> {
  addon: Addon<Server>;
  insertionIndex: number;
}

export class Registry<Server = unknown> {
  private entries: RegistryEntry<Server>[] = [];
  private counter = 0;

  register(addon: Addon<Server>): this {
    if (!addon || typeof addon.name !== "string" || typeof addon.run !== "function") {
      throw new TypeError("Addon must be { name: string, priority: number, run: function }");
    }
    this.entries.push({ addon, insertionIndex: this.counter++ });
    return this;
  }

  list(): Addon<Server>[] {
    return [...this.entries]
      .sort((a, b) => {
        if (a.addon.priority !== b.addon.priority) return a.addon.priority - b.addon.priority;
        return a.insertionIndex - b.insertionIndex;
      })
      .map((e) => e.addon);
  }

  async runAll(
    server: Server,
    config: ResolvedBootstrapConfig,
    ctx: RuntimeContext<Server>,
  ): Promise<LoaderReports> {
    const reports: LoaderReports = {};
    for (const addon of this.list()) {
      try {
        const report = await addon.run(server, config, ctx);
        reports[addon.name] = report;
      } catch (err) {
        const fallback = createLoaderReport(addon.name);
        reportError(fallback, "run", err);
        reports[addon.name] = fallback;
      }
    }
    return reports;
  }
}

export function composeAddons<S>(...groups: (Addon<S>[] | Addon<S>)[]): Addon<S>[] {
  const out: Addon<S>[] = [];
  for (const g of groups) {
    if (Array.isArray(g)) out.push(...g);
    else out.push(g);
  }
  return out;
}
